src/main/java/org/mskcc/cmo/metadb/service/impl/CmoLabelGeneratorServiceImpl.java

```diff
@@ -0,0 +1,251 @@
+    // example: C-1235-X001-d
+    public static final Pattern CMO_SAMPLE_ID_REGEX =
+            Pattern.compile("^C-([a-zA-Z0-9]+)-([NTRMLUPSGX])([0-9]{3})-([dr])$");
+    // example: JH123-12345T
+    public static final Pattern CMO_CELLLINE_ID_REGEX =
+            Pattern.compile("^([A-Za-z0-9]+)-([A-Za-z0-9]+)$");
+    public static final String CMO_LABEL_SEPARATOR = "-";
+
+    @Autowired
+    private MetadbRestService metadbRestService;
+
+    // globals for mapping sample type abbreviations
+    private static Map<SpecimenType, String> SPECIMEN_TYPE_ABBREV_MAP = new HashMap<>();
+    private static Map<SampleOrigin, String> SAMPLE_ORIGIN_ABBREV_MAP = new HashMap<>();
+    private static Map<CmoSampleClass, String> SAMPLE_CLASS_ABBREV_MAP = new HashMap<>();
+    private static final List<SampleOrigin> KNOWN_CFDNA_SAMPLE_ORIGINS =
+            Arrays.asList(SampleOrigin.URINE,
+                    SampleOrigin.CEREBROSPINAL_FLUID,
+                    SampleOrigin.PLASMA,
+                    SampleOrigin.WHOLE_BLOOD);
+    private static final String SAMPLE_ORIGIN_ABBREV_DEFAULT = "T";
+
+    @Autowired
+    private void initSampleTypeAbbreviationMapping() {
+        // set up mapping by specimen type
+        SPECIMEN_TYPE_ABBREV_MAP.put(SpecimenType.PDX, "X");
+        SPECIMEN_TYPE_ABBREV_MAP.put(SpecimenType.XENOGRAFT, "X");
+        SPECIMEN_TYPE_ABBREV_MAP.put(SpecimenType.XENOGRAFTDERIVEDCELLLINE, "X");
+        SPECIMEN_TYPE_ABBREV_MAP.put(SpecimenType.ORGANOID, "G");
+
+        // set up mapping by sample origin
+        SAMPLE_ORIGIN_ABBREV_MAP.put(SampleOrigin.URINE, "U");
+        SAMPLE_ORIGIN_ABBREV_MAP.put(SampleOrigin.CEREBROSPINAL_FLUID, "S");
+        SAMPLE_ORIGIN_ABBREV_MAP.put(SampleOrigin.PLASMA, "L");
+        SAMPLE_ORIGIN_ABBREV_MAP.put(SampleOrigin.WHOLE_BLOOD, "L");
+
+        // set up mapping by sample class
+        SAMPLE_CLASS_ABBREV_MAP.put(CmoSampleClass.UNKNOWN_TUMOR, "T");
+        SAMPLE_CLASS_ABBREV_MAP.put(CmoSampleClass.LOCAL_RECURRENCE, "R");
+        SAMPLE_CLASS_ABBREV_MAP.put(CmoSampleClass.PRIMARY, "P");
+        SAMPLE_CLASS_ABBREV_MAP.put(CmoSampleClass.RECURRENCE, "R");
+        SAMPLE_CLASS_ABBREV_MAP.put(CmoSampleClass.METASTASIS, "M");
+        SAMPLE_CLASS_ABBREV_MAP.put(CmoSampleClass.NORMAL, "N");
+        SAMPLE_CLASS_ABBREV_MAP.put(CmoSampleClass.ADJACENT_NORMAL, "N");
+        SAMPLE_CLASS_ABBREV_MAP.put(CmoSampleClass.ADJACENT_TISSUE, "T");
+    }
+
+    @Override
+    public String generateCmoSampleLabel(SampleMetadata sampleMetadata) {
+        // if sample is a cellline sample then generate a cmo cellline label
+        if (isCmoCelllineSample(sampleMetadata)) {
+            return generateCmoCelllineSampleLabel(sampleMetadata);
+        }
+
+        List<SampleMetadata> existingSamples = new ArrayList<>();
+        try {
+            existingSamples =
+                    metadbRestService.getSampleMetadataListByCmoPatientId(sampleMetadata.getCmoPatientId());
```

Comment:
Can getCmoPatientId return null or is this checked by upstream validator?

Reply:
Checked by the validator